File: CS/XtraTabWithButtons/ViewInfo/TabButtonsPanelDescendant.cs
using System;
using System.Collections.Generic;
using System.Linq;
using DevExpress.XtraTab.Buttons;
using DevExpress.XtraTab.ViewInfo;
using DevExpress.XtraTab;
using DevExpress.Utils;
using System.Drawing;

namespace XtraTabWithButtons
{
    public class TabButtonsPanelDescendant : TabButtonsPanel
    {
        private XtraTabPageDescendant _PageDescendant;

        public TabButtonsPanelDescendant(BaseTabControlViewInfo tabViewInfo, XtraTabPageDescendant pagedes)
            : base(tabViewInfo)
        {
            _PageDescendant = pagedes;
        }

        public XtraTabPageDescendant PageDescendant
        {
            get { return _PageDescendant; }
        }

        protected override void CreateButtonsCore(TabButtons buttons, CustomHeaderButtonCollection userButtons)
        {
            base.CreateButtonsCore(buttons, PageDescendant.CustomButtons);
        }

        protected override void OnClickButton(TabButtonInfo button)
        {
            keepButtons++;
            try
            {
                if (CanRaiseHeaderButtonClick(button))
                    PageDescendant.RaiseCustomPageButtonClick(button);
            }
            finally
            {
                keepButtons--;
            }
            if (!TabViewInfo.IsDisposing)
                TabViewInfo.LayoutChanged();
        }

        public ToolTipControlInfo GetToolTip(Point point)
        {
            return GetToolTipInfo(point);
        }
    }
}

File: VB/XtraTabWithButtons/ViewInfo/TabButtonsPanelDescendant.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Collections.Generic
Imports System.Linq
Imports DevExpress.XtraTab.Buttons
Imports DevExpress.XtraTab.ViewInfo
Imports DevExpress.XtraTab
Imports DevExpress.Utils
Imports System.Drawing

Namespace XtraTabWithButtons
	Public Class TabButtonsPanelDescendant
		Inherits TabButtonsPanel
		Private _PageDescendant As XtraTabPageDescendant

		Public Sub New(ByVal tabViewInfo As BaseTabControlViewInfo, ByVal pagedes As XtraTabPageDescendant)
			MyBase.New(tabViewInfo)
			_PageDescendant = pagedes
		End Sub

		Public ReadOnly Property PageDescendant() As XtraTabPageDescendant
			Get
				Return _PageDescendant
			End Get
		End Property

		Protected Overrides Sub CreateButtonsCore(ByVal buttons As TabButtons, ByVal userButtons As CustomHeaderButtonCollection)
			MyBase.CreateButtonsCore(buttons, PageDescendant.CustomButtons)
		End Sub

		Protected Overrides Sub OnClickButton(ByVal button As TabButtonInfo)
			keepButtons += 1
			Try
				If CanRaiseHeaderButtonClick(button) Then
					PageDescendant.RaiseCustomPageButtonClick(button)
				End If
			Finally
				keepButtons -= 1
			End Try
			If (Not TabViewInfo.IsDisposing) Then
				TabViewInfo.LayoutChanged()
			End If
		End Sub

		Public Function GetToolTip(ByVal point As Point) As ToolTipControlInfo
			Return GetToolTipInfo(point)
		End Function
	End Class
End Namespace
